SCons/Util.py

```diff
@@ -1468,66 +1468,155 @@ def f(self, x, y):
-        :return: String of Hex digits representing the signature
-        """
-        m = hashlib.md5()
-        with open(fname, "rb") as f:
-            while True:
-                blck = f.read(chunksize)
-                if not blck:
-                    break
-                m.update(to_bytes(blck))
-        return m.hexdigest()
-else:
-    # if md5 algorithm not available, just return data unmodified
-    # could add alternative signature scheme here
-    md5 = False
+    :param hash_format: Hash format to use.
+    :return: hashlib object.
+    """
+    if hash_format is None:
+        if _hash_function is None:
+            raise Exception('There is no default hash function. Did you call '
+                            'a hashing function before SCons was initialized?')
+        return _hash_function()
+    elif not hasattr(hashlib, hash_format):
+        raise Exception(
+            'Hash format "%s" is not available in your Python interpreter.' %
+            hash_format)
+    else:
+        return getattr(hashlib, hash_format)()
 
-    def MD5signature(s):
-        return str(s)
 
-    def MD5filesignature(fname, chunksize=65536):
-        with open(fname, "rb") as f:
-            result = f.read()
-        return result
+def hash_signature(s, hash_format=None):
+    """
+    Generate hash signature of a string
 
+    :param s: either string or bytes. Normally should be bytes
+    :param hash_format: Specify to override default hash format
+    :return: String of hex digits representing the signature
+    """
+    m = _get_hash_object(hash_format)
+    try:
+        m.update(to_bytes(s))
+    except TypeError:
+        m.update(to_bytes(str(s)))
 
-def MD5collect(signatures):
+    return m.hexdigest()
+
+
+def hash_file_signature(fname, chunksize=65536, hash_format=None):
+    """
+    Generate the md5 signature of a file
+
+    :param fname: file to hash
+    :param chunksize: chunk size to read
+    :param hash_format: Specify to override default hash format
+    :return: String of Hex digits representing the signature
+    """
+    m = _get_hash_object(hash_format)
+    with open(fname, "rb") as f:
+        while True:
+            blck = f.read(chunksize)
+            if not blck:
+                break
+            m.update(to_bytes(blck))
+    return m.hexdigest()
+
+
+def hash_collect(signatures, hash_format=None):
     """
     Collects a list of signatures into an aggregate signature.
 
-    signatures - a list of signatures
-    returns - the aggregate signature
+    :param signatures: a list of signatures
+    :param hash_format: Specify to override default hash format
+    :return: - the aggregate signature
     """
     if len(signatures) == 1:
         return signatures[0]
     else:
-        return MD5signature(', '.join(signatures))
+        return hash_signature(', '.join(signatures), hash_format)
+
+
+def MD5signature(s):
+    """
+    Deprecated. Use hash_signature instead.
+    """
+    return hash_signature(s)
```

Comment:
Just a placeholder. We may want to add some sort of deprecation notification on these calls.